File: decode_demo/fanyi.youdao.py
import execjs
import requests

word = input("请输入要翻译的单词：")
# 1. 获取翻译结果

cookies = {
    'JSESSIONID': 'abckWzgSNk692fxiW7Xkz',
    'OUTFOX_SEARCH_USER_ID_NCOO': '1879393198.1356268',
    'OUTFOX_SEARCH_USER_ID': '-340494695@112.87.207.61',
    '_uetsid': '83f54070e52311efb5e0f56e4406b0ee',
    '_uetvid': 'f7ec2d60d47811efb06675be584fef08',
    'DICT_DOCTRANS_SESSION_ID': 'YWM0YWNjYjItNmJlMS00MDRkLWFlOGUtMGVkOWY0MDYxMjMy',
}

headers = {
    'Accept': 'application/json, text/plain, */*',
    'Accept-Language': 'zh-CN,zh;q=0.9,en;q=0.8,en-GB;q=0.7,en-US;q=0.6',
    'Connection': 'keep-alive',
    'Content-Type': 'application/x-www-form-urlencoded',
    # 'Cookie': 'JSESSIONID=abckWzgSNk692fxiW7Xkz; OUTFOX_SEARCH_USER_ID_NCOO=1879393198.1356268; OUTFOX_SEARCH_USER_ID=-340494695@112.87.207.61; _uetsid=83f54070e52311efb5e0f56e4406b0ee; _uetvid=f7ec2d60d47811efb06675be584fef08; DICT_DOCTRANS_SESSION_ID=YWM0YWNjYjItNmJlMS00MDRkLWFlOGUtMGVkOWY0MDYxMjMy',
    'Origin': 'https://fanyi.youdao.com',
    'Referer': 'https://fanyi.youdao.com/',
    'Sec-Fetch-Dest': 'empty',
    'Sec-Fetch-Mode': 'cors',
    'Sec-Fetch-Site': 'same-site',
    'User-Agent': 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/132.0.0.0 Safari/537.36 Edg/132.0.0.0',
    'sec-ch-ua': '"Not A(Brand";v="8", "Chromium";v="132", "Microsoft Edge";v="132"',
    'sec-ch-ua-mobile': '?0',
    'sec-ch-ua-platform': '"Windows"',
}

data = {
    'i': word,
    'from': 'auto',
    'to': '',
    'useTerm': 'false',
    'dictResult': 'true',
    'keyid': 'webfanyi',
    'sign': 'b95245d5b1fddeba5937666d596e5efb',
    'client': 'fanyideskweb',
    'product': 'webfanyi',
    'appVersion': '1.0.0',
    'vendor': 'web',
    'pointParam': 'client,mysticTime,product',
    'mysticTime': '1738912668484',
    'keyfrom': 'fanyi.web',
    'mid': '1',
    'screen': '1',
    'model': '1',
    'network': 'wifi',
    'abtest': '0',
    'yduuid': 'abcdefg',
}

with open('fanyi.youdao.js', 'r', encoding='utf-8') as f:
    js_code = f.read()

ctx = execjs.compile(js_code)

params = ctx.call('getParams')

data['sign'] = params['sign']
data['mysticTime'] = params['mysticTime']

response = requests.post('https://dict.youdao.com/webtranslate', cookies=cookies, headers=headers, data=data)

encrypted_data = response.text

# 2. 解密翻译结果
decoded_data = ctx.call('decode', encrypted_data)

print(decoded_data)

File: decode_demo/fanyi.youdao.js
const crypto = require('crypto');
const {Buffer} = require('buffer');

d = "fanyideskweb"
u = "webfanyi"

// e是从另一个请求获得, 但好像是个固定值
// https://dict.youdao.com/webtranslate/key?keyid=webfanyi-key-getter&sign=a0f7514bec04b363bcdbe921d5ed22d8&client=fanyideskweb&product=webfanyi&appVersion=1.0.0&vendor=web&pointParam=client,mysticTime,product&mysticTime=1738990548447&keyfrom=fanyi.web&mid=1&screen=1&model=1&network=wifi&abtest=0&yduuid=abcdefg
e = "Vy4EQ1uwPkUoqvcP1nIu6WiAjxFeA3Y9"

function S(e, t) {
    return _(`client=${d}&mysticTime=${e}&product=${u}&key=${t}`)
}

function _(e) {
    return md5(e.toString());
}

function md5(e) {
    return crypto.createHash('md5').update(e).digest('hex');
}

function getParams() {
    params = {}

    a = (new Date).getTime()
    params.mysticTime = a
    params.sign = S(a, e)

    return params
}

console.log(getParams())

o = "Z21kD9ZK1ke6ugku2ccWu4n6eLnvoDT0YgGi0y3g-v0B9sYqg8L9D6UERNozYOHqoQzNRUCyI06c5-fXHEG36QL8otVzFy9yH2bX0qRjbBJpLVuhShQ1gW2_ibzi_ZCov8kB9YPV40qhWUW6UZjEgztaeTrgclA9HMQRe8jH_dIiX429nANGhMRtXGkwVsVx8WwuH1xZpEOtb-sNI_c962nsPfhvyt6w7rAd4IVPv2ue4_1DykKKPi4muiSsWWwPBKI7hmO9MrVBnj2RAwk2A0OgL_FLa_vgLMutRioYbU0LG8TvuqiJb9yMHrNNg-ibYzWHokVYUdZ7pfKY8ory-kwXa1RI-Vl_9Z1-zQfxgfTAl8kkMztxGicK2akHcHI2xOkBKP8ZUa4unNKRueBrOWM8vbQxLTZoqpBF3y7MdwmrA3G2JkWmSpmRSWBU9jC8HUeObnX5vC7RiJ1tGOKRg2ltxCCzrGbJD-49M7CLe-B6Q4ndpsqGz6276HOiwp8unL4rUQf-vbp_8CrE6aTENWariAAoupAsNXP4fgly6v0zdoO8xWFkthrrRjScI8RyS21_Z4nyngLjKfadc11JGGtNb65BzWiK8qFQTt1Zl3hYnQfKFzgrUYe0CJmNg9umBjKaEa6yt8XVNHJPuJffFAyTgwgU0m06mfU8kx_Ka7SG4emtM4PryQEtQlP3R0gTGun495UORo9KZk19utHnrfrSLsqeX-JglRDs6_75Yv4cnHne06oPRYhFjh_Q35Kps-BVhaXfKAeGLc2WnS0Ij3Tq24sU2taVojTdAQSB4tpPFA3vcHdu7p2CLCRkZc25ArFM6oTsjffagAK8rqRs-Y34p_tohegFf5O1wv3LDFUqlRi3tgWmjxyRz0y9JyYefjBi-5_tWhgLmC-wc-AzoOyYNZ5J7ke9Cphlp3FLDBsAgW24fOhsBdV7DS3kpE6kS4AZy-z3Tv3fD_1bso5w1MZYdRHSUxr2SGjvkyKSd_RXox_tZGXgkRhLJllE626cqITks5jNPwF_0t90P-wkxv3cdL_h2G4OgxXTTr16VLIrLp5M6HC5GySz2DgNz56t0twoXmC1OxGLauSh0y2TWPGyB0Rx5N1rH0B4Q3sdlxYtuu9RO4I407dlAgdFpbxfl0WK3XBj1erfc3w5oKY99ERCjCe7QdJZ-K8OAK0vjQWlGSrLoAxwjYAcSkVbBaDEmKaOM6eGzDRdpbHjn-s8rA=="

// 这两个可能也是固定值
decodeKey = "ydsecret://query/key/B*RGygVywfNBwpmBaZg*WT7SIOUP2T0C9WHMZN39j^DAdaZhAnxvGcCY6VYFwnHl"
decodeIv = "ydsecret://query/iv/C@lZe2YzHtZ2CYgaXKSVfsb7Y4QWHjITPPZ0nQp87fBeJ!Iv6v^6fvi2WN@bYpJ4"

function T(e) {
    return crypto.createHash("md5").update(e).digest()
}

function decode(e) {
    if (!e)
        return null;
    const o = Buffer.alloc(16, T(decodeKey))
        , n = Buffer.alloc(16, T(decodeIv))
        , r = crypto.createDecipheriv("aes-128-cbc", o, n);
    let s = r.update(e, "base64", "utf-8");
    return s += r.final("utf-8"),
        s
}

console.log(decode(o))






